mypy/build.py

```diff
@@ -812,202 +820,122 @@ def remove_cwd_prefix_from_path(p: str) -> str:
     return p
 
 
-# Cache find_module: (id, lib_path) -> result.
-find_module_cache = {}  # type: Dict[Tuple[str, Tuple[str, ...]], Optional[str]]
-
-# Cache some repeated work within distinct find_module calls: finding which
-# elements of lib_path have even the subdirectory they'd need for the module
-# to exist.  This is shared among different module ids when they differ only
-# in the last component.
-find_module_dir_cache = {}  # type: Dict[Tuple[str, Tuple[str, ...]], List[str]]
-
-# Cache directory listings.  We assume that while one os.listdir()
-# call may be more expensive than one os.stat() call, a small number
-# of os.stat() calls is quickly more expensive than caching the
-# os.listdir() outcome, and the advantage of the latter is that it
-# gives us the case-correct filename on Windows and Mac.
-find_module_listdir_cache = {}  # type: Dict[str, Optional[List[str]]]
-
-# Cache for is_file()
-find_module_is_file_cache = {}  # type: Dict[str, bool]
-
-# Cache for isdir(join(head, tail))
-find_module_isdir_cache = {}  # type: Dict[Tuple[str, str], bool]
-
-
-def find_module_clear_caches() -> None:
-    find_module_cache.clear()
-    find_module_dir_cache.clear()
-    find_module_listdir_cache.clear()
-    find_module_is_file_cache.clear()
-    find_module_isdir_cache.clear()
-
+class FindModuleCache:
+    """Module finder with integrated cache.
 
-def list_dir(path: str) -> Optional[List[str]]:
-    """Return a cached directory listing.
+       Module locations and some intermediate results are cached internally
```

Comment:
Style nit: Indents seem off unless github diff view is acting up. Just to be extra clear, indents should be like this:

```py
class Foo:
    """First line.

    Second line (4 space indent).
    """
```